src/main.rs

```diff
@@ -70,27 +79,94 @@ async fn main() {
-
-    let redirect_url = env::var("REDIRECT_URL").unwrap_or_else(|_| OAUTH_REDIRECT_URL.to_string());
-    let auth_url = env::var("AUTH_URL").unwrap_or_else(|_| OAUTH_AUTH_URL.to_string());
-    let token_url = env::var("TOKEN_URL").unwrap_or_else(|_| OAUTH_TOKEN_URL.to_string());
-
-    BasicClient::new(
-        ClientId::new(client_id),
-        Some(ClientSecret::new(client_secret)),
-        AuthUrl::new(auth_url).unwrap(),
-        Some(TokenUrl::new(token_url).unwrap()),
-    )
-    .set_redirect_uri(RedirectUrl::new(redirect_url).unwrap())
+#[derive(Clone)]
+struct SiweOAuthClient {
+    client: BasicClient,
+}
+
+impl Deref for SiweOAuthClient {
+    type Target = BasicClient;
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
+}
+
+fn siwe_oauth_client() -> SiweOAuthClient {
+    let client_id = env::var("SIWE_CLIENT_ID").expect("Missing SIWE_CLIENT_ID!");
+    let client_secret = env::var("SIWE_CLIENT_SECRET").expect("Missing SIWE_CLIENT_SECRET!");
+
+    let redirect_url =
+        env::var("SIWE_REDIRECT_URL").unwrap_or_else(|_| SIWE_OAUTH_REDIRECT_URL.to_string());
+    let auth_url = env::var("SIWE_AUTH_URL").unwrap_or_else(|_| SIWE_OAUTH_AUTH_URL.to_string());
+    let token_url = env::var("SIWE_TOKEN_URL").unwrap_or_else(|_| SIWE_OAUTH_TOKEN_URL.to_string());
+
+    SiweOAuthClient {
+        client: BasicClient::new(
+            ClientId::new(client_id),
+            Some(ClientSecret::new(client_secret)),
+            AuthUrl::new(auth_url).unwrap(),
+            Some(TokenUrl::new(token_url).unwrap()),
+        )
+        .set_redirect_uri(RedirectUrl::new(redirect_url).unwrap()),
+    }
+}
+
+#[derive(Clone)]
+struct GithubOAuthClient {
+    client: BasicClient,
+}
+
+impl Deref for GithubOAuthClient {
+    type Target = BasicClient;
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
+}
+
+fn github_oauth_client() -> GithubOAuthClient {
+    let client_id = env::var("GITHUB_CLIENT_ID").expect("Missing GITHUB_CLIENT_ID!");
+    let client_secret = env::var("GITHUB_CLIENT_SECRET").expect("Missing GITHUB_CLIENT_SECRET!");
+    let redirect_url =
+        env::var("GITHUB_REDIRECT_URL").unwrap_or_else(|_| GITHUB_OAUTH_REDIRECT_URL.to_string());
+    let auth_url =
+        env::var("GITHUB_AUTH_URL").unwrap_or_else(|_| GITHUB_OAUTH_AUTH_URL.to_string());
+    let token_url =
+        env::var("GITHUB_TOKEN_URL").unwrap_or_else(|_| GITHUB_OAUTH_TOKEN_URL.to_string());
+    GithubOAuthClient {
+        client: BasicClient::new(
+            ClientId::new(client_id),
+            Some(ClientSecret::new(client_secret)),
+            AuthUrl::new(auth_url).unwrap(),
+            Some(TokenUrl::new(token_url).unwrap()),
+        )
+        .set_redirect_uri(RedirectUrl::new(redirect_url).unwrap()),
+    }
 }
 
 async fn hello_world() -> Html<&'static str> {
     Html("<h1>Server is Running</h1>")
 }
 
+#[derive(Clone)]
+pub(crate) struct AppConfig {
```

Comment:
Can you put these into the constants directory?

Reply:
Do you mean "move the `AppConfig` type to the constants file" or "move the fields to the constants file and get rid of `AppConfig`"? If the first one, then sure. If the second: I'm introducing `AppConfig` because at some point I'll want to write automated tests for this, and then I'll want to stub the external apis / swap constants around, which is why I'm specifically making them injectable and not constant :)

Reply:
Yeah the second, I think this can be done without AppConfig being an Extension, since we don't really care about the parameters from each auth provider, just that some validation function passes when it is given a user profile as an argument. 

Its not blocking for me though and we can pick this up, when the automated tests are added!